### PR TITLE
Send IPv4 GARP and IPv6 Unsolicited NA in "cmdAdd"

### PR DESCRIPTION
**Pull request by wizhaoredhat**
In "cmdAdd", SRIOV-CNI would construct and send IPv4 Gratuitous ARP
and/or Unsolicited Neighbor Advertisement depending on the IP
addresses configured by IPAM. The reason why this change is needed
is for the scenario when an IP address is reused by IPAM with
different interfaces (with different link-layer addresses). This can
occur when pods are deleted and created. For performance reasons,
sending of GARP and/or Unsolicited NA would update invalid ARP/Neighbor
caches in other neighbors/nodes.

Also we set IPv4 ARP Notify and IPv6 Neighbor Discovery Notify in sysfs
for each interface. This will send GARP and/or Unsolicited NA when the
interface is either brought up or the link-layer address changes. This
is useful in cases where an application reenables the interface or the
MAC address configuration is changed.

Some new packages were added, thus go.mod and go.sum were modified
accordingly.

Mocked PciUtils for sriov tests since sriov.go would call PciUtils to
set IPv4 ARP Notify and IPv6 Neighbor Discovery.

Fixes k8snetworkplumbingwg#177

Signed-off-by: William Zhao <wizhao@redhat.com>

### CODE REVIEW

**Comment by wizhaoredhat** [comment]
@adrianchiris @martinkennelly @zshi-redhat  @bn222 @Eoghan1232 : When you have an opportunity it would be much appreciated if you reviewed these changes in my PR.

**Comment by wizhaoredhat** [comment]
I believe there was an intermittent problem with checking out some golang libraries. I hope that is what caused the kubernetes ci to fail. Could someone help with reruning the workflow?

File: pkg/utils/utils.go
Comment (adrianchiris) [inline pkg/utils/utils.go:302]: I still need to give this PR a deeper look.


i see : https://github.com/mdlayher/arp implements sending arp.
perhaps we can use here ? 


in any case i feel we should put the logic which builds the packets  (for both arp and and NA) in a separate file 

e.g `utils/packet.go`

File: pkg/utils/utils.go
Comment (adrianchiris) [inline pkg/utils/utils.go:50]: if ipv6 is disabled for interface, will the path exist ? 

e.g
`sudo sysctl -w net.ipv6.conf.eth0.disable_ipv6=1`
`net.ipv6.conf.all.disable_ipv6=1`

File: pkg/sriov/sriov.go
Comment (adrianchiris) [inline pkg/sriov/sriov.go:202]: should we fail here ?

maybe it should be more like:

`tryEnableArpAndNdiscNotify()`

File: pkg/utils/utils.go
Comment (adrianchiris) [reply pkg/utils/utils.go:50]: it does :)

File: pkg/utils/utils.go
Comment (s1061123) [inline pkg/utils/utils.go:362]: If you send unsolicited NA, then this message changes existing entry to STALE as RFC Section 7.2.5's following statement. Does it make sense? 

```
   - If the advertisement's Solicited flag is set, the state of the
     entry is set to REACHABLE; otherwise, it is set to STALE.
```

Comment (wizhaoredhat) [reply pkg/utils/utils.go:362]: In [7.2.6](https://www.rfc-editor.org/rfc/rfc4861.html#section-7.2.6).  Sending Unsolicited Neighbor Advertisements:

"In either case, neighboring nodes will immediately change the state of their Neighbor Cache entries for the Target Address to STALE, prompting them to verify the path for reachability. "

I believe IPv6 is special in a way that it tests connectivity both ways, that's why it needs to verify the path for reachability.

File: pkg/sriov/sriov.go
Comment (wizhaoredhat) [reply pkg/sriov/sriov.go:202]: Is there a concern that this function would fail in different environments? Or is there a scenario that you are thinking that would fail, but only fails because the sysfs variables aren't supported?

File: pkg/utils/utils.go
Comment (wizhaoredhat) [reply pkg/utils/utils.go:50]: Thanks for confirming that they exist in the case when ipv6 is disable :) !

File: pkg/utils/utils.go
Comment (wizhaoredhat) [reply pkg/utils/utils.go:302]: I took a look at the implementation of the library. I don't believe it supports sending gratuitous ARPs. Rather it is a library for ARP resolution. This library supports gratuitous ARPs: https://github.com/j-keck/arping . However I am not in favor of it's implementation since it uses raw sockets and builds the entire Ethernet Frame. (https://github.com/j-keck/arping/blob/c50f4319d307d95d1e2b344c4504158705e919de/arping_linux.go#L21)

**Review by SchSeba — COMMENTED** [review]
The PR looks good thanks for the contribution!

I have a general question if we change the sys/net file and then change the nic to up the kernel should send the request for us no?

base on (https://sysctl-explorer.net/net/ipv6/ndisc_notify/)

File: pkg/utils/utils.go
Comment (bn222) [inline pkg/utils/utils.go:315]: Let's not eat up all the errors, but instead propagate them.

File: pkg/utils/utils.go
Comment (bn222) [reply pkg/utils/utils.go:302]: I looked at the arping library that you mentioned. At the API level, it doesn't seem to expose that it builds the full packet including ethernet. I'm in favor of still relying on that external library to build the packet, or patching that library to (add an option to) not use raw sockets.

In any case, it feels like we are just dumping "a small arp library" into `pkg/utils/utils.go` where we are already doing all kinds of other things.

File: pkg/utils/utils.go
Comment (wizhaoredhat) [reply pkg/utils/utils.go]: I was thinking that it would never error because they are fixed values. But I agree with you, I think its is better to return the errors properly just in case. This is in my latest commit.

File: pkg/utils/utils.go
Comment (wizhaoredhat) [reply pkg/utils/utils.go]: I put all the logic for sending packets (ARPs/Unsolicited NA) in a separate file. I hope this works for everyone.

**Comment by wizhaoredhat** [comment]
> The PR looks good thanks for the contribution!
> 
> I have a general question if we change the sys/net file and then change the nic to up the kernel should send the request for us no?
> 
> base on (https://sysctl-explorer.net/net/ipv6/ndisc_notify/)

Glad to be able to contribute!

To your question, yes that is correct. This is the method that Zenghui was first prototyping. However after some debugging and testing. We found some issues:
1) The interface must be up for IPAM to function correctly. E.g. dhcp IPAM.
(Thus we have to down the interface after IPAM config and bring it back up)
2) IP addresses can be lost or changed. E.g By default sysctl "keep_addr_on_down" for IPv6 is set to zero. This means that if the interface is "downed", the IPv6 address(es) configured statically are lost.
3) Also flapping a interface link when the majority of the network configuration was completed by SRIOV-CNI and IPAM to send a gratuitous ARP or unsolicited NA seemed a bit hacky in my opinion.

Constructing the packets manually and sending them seemed like the correct path to go.

File: pkg/utils/utils.go
Comment (s1061123) [reply pkg/utils/utils.go]: So, regarding ipv6, this change is to invalidate current existing entry from the node. Is it right?

Comment (wizhaoredhat) [reply pkg/utils/utils.go]: Yes that's the most relevant aspect of this change where there could be a reuse of an IP address with a different pod (different MAC address on the container interface).

File: cmd/sriov/main.go
Comment (bn222) [inline cmd/sriov/main.go:171]: Let's make this a function

File: pkg/sriov/sriov.go
Comment (wizhaoredhat) [reply pkg/sriov/sriov.go:202]: I could only think of 2 scenarios where calling this function would fail:
1) Not enough permissions to set the sysctl variable
2) The podifName is incorrect.
However both of these scenarios would be considered major malfunctions and would require us to log an error.
@adrianchiris

File: pkg/utils/utils.go
Comment (wizhaoredhat) [reply pkg/utils/utils.go]: I hope I was able to answer your concerns.

File: pkg/sriov/sriov.go
Comment (adrianchiris) [reply pkg/sriov/sriov.go:202]: so, my question is, whether we regard this as an optimization or a mandatory operation.

for the former we should not fail. for the latter we should.

Comment (adrianchiris) [reply pkg/sriov/sriov.go:202]: according to this PR cover letter:

```
For performance reasons,
sending of GARP and/or Unsolicited NA would update invalid ARP/Neighbor
caches in other neighbors/nodes.
```

its the former

Comment (wizhaoredhat) [reply pkg/sriov/sriov.go:202]: @adrianchiris 

I see what you mean. There are 2 roadblocks I have:
1) SRIOV-CNI is a Multus plugin and thus doesn't have the ability to log warnings/informational. We can only report errors, but this in turn causes the pods to fail to create.
2) Yes, this is an optimization. But the causes for this function to fail would be abnormal. There would be no reason for a healthy system to fail calling these functions.

Are we ok with running the function and if it fails, we don't log anything? Otherwise we need a new logging feature for SRIOV-CNI.

File: cmd/sriov/main.go
Comment (wizhaoredhat) [reply cmd/sriov/main.go:171]: Good catch, done and made other changes too.

File: pkg/utils/packet.go
Comment (bn222) [inline pkg/utils/packet.go:41]: Why return (...)? why not return ... 

Same comment for others.

File: pkg/sriov/sriov.go
Comment (bn222) [reply pkg/sriov/sriov.go:202]: Under the premise that this is strictly an optimization (i.e. there are no other hard-to-debug side-effects of not having this in place), I don't mind not logging anything.

Comment (wizhaoredhat) [reply pkg/sriov/sriov.go:202]: @adrianchiris 

In the spirit of moving this PR closer to being merged. I agree that it is an optimization, and there shouldn't be any hard to debug side effects. I am ok with removing any logging and just throwing away the error message. Is this what you had in mind?

**Comment by wizhaoredhat** [comment]
@adrianchiris @zshi-redhat @bn222 @s1061123 @SchSeba : Please review again at your convenience. I would like to have this feature in soon. Thanks!

File: pkg/utils/packet.go
Comment (wizhaoredhat) [reply pkg/utils/packet.go:41]: Just a personal preference for readability. I can make this minor change.

Comment (wizhaoredhat) [reply pkg/utils/packet.go]: Made the change.

File: pkg/sriov/sriov.go
Comment (wizhaoredhat) [reply pkg/sriov/sriov.go]: I made the change to not fail with error.

File: pkg/utils/packet.go
Comment (bn222) [inline pkg/utils/packet.go:41]: == has higher precedence than &&. Other parts of the code also don't use explicit () in this case.

Comment (wizhaoredhat) [reply pkg/utils/packet.go]: Just a personal preference for readability. But made the changes to follow the existing code style.

File: pkg/utils/packet.go
Comment (adrianchiris) [inline pkg/utils/packet.go:30]: WDYT about moving IsValidMacAddress, IsIPv4 and IsIPv6 to utils.go ?

keeping here just the bits which deal solely with sending GARP/NA packets

**Review by adrianchiris — COMMENTED** [review]
Thanks for working on this @wizhaoredhat  !

apologies for the late review.

File: cmd/sriov/main.go
Comment (wizhaoredhat) [reply cmd/sriov/main.go]: Done. Switched over to using netlink.

File: cmd/sriov/main.go
Comment (wizhaoredhat) [reply cmd/sriov/main.go]: Very good point. I moved the logic up.

File: cmd/sriov/main.go
Comment (wizhaoredhat) [reply cmd/sriov/main.go]: Done.

File: cmd/sriov/main.go
Comment (wizhaoredhat) [reply cmd/sriov/main.go]: Done.

File: cmd/sriov/main.go
Comment (wizhaoredhat) [reply cmd/sriov/main.go]: Done.

File: pkg/sriov/sriov_test.go
Comment (wizhaoredhat) [reply pkg/sriov/sriov_test.go]: I created a new folder and places the mock files there.

File: pkg/utils/packet.go
Comment (wizhaoredhat) [reply pkg/utils/packet.go]: I think that's a good idea. Made the change.

Comment (wizhaoredhat) [reply pkg/utils/packet.go]: Thanks for reviewing my changes. Much appreciated!

File: pkg/utils/utils.go
Comment (wizhaoredhat) [reply pkg/utils/utils.go]: Done.

File: cmd/sriov/main.go
Comment (wizhaoredhat) [reply cmd/sriov/main.go]: Done.

**Comment by bn222** [comment]
The commit has become quite large. Would it be possible to split it up a bit?

**Comment by SchSeba** [comment]
For the CI issue I am an open PR for it 

https://github.com/k8snetworkplumbingwg/sriov-cni/pull/223

File: pkg/utils/utils.go
Comment (SchSeba) [inline pkg/utils/utils.go:292]: should we also add a regex validation here?

File: pkg/sriov/sriov.go
Comment (SchSeba) [inline pkg/sriov/sriov.go:112]: question should we return error here and continue to skip the error on `announceIPs`?

File: pkg/utils/utils.go
Comment (adrianchiris) [reply pkg/utils/utils.go:292]: net.HardwareAddr tyoe is just a list of bytes.
maybe just check its not all zeroes ?

File: pkg/sriov/sriov.go
Comment (adrianchiris) [reply pkg/sriov/sriov.go:112]: @SchSeba this is regarded as performance enhancement and so i would not want it to fail if not needed.

File: cmd/sriov/main.go
Comment (adrianchiris) [reply cmd/sriov/main.go:44]: also lets add a docstring for this one

**Review by adrianchiris — COMMENTED** [review]
Overall looks good, once @SchSeba comments are addressed and PR rebased im LGTM

File: pkg/sriov/sriov.go
Comment (wizhaoredhat) [reply pkg/sriov/sriov.go:112]: Yes we agreed that any function that is part of any performance enhancement shall not fail. Resolving this comment.

File: pkg/utils/utils.go
Comment (wizhaoredhat) [reply pkg/utils/utils.go]: I added a check for zeros and broadcast. I hope this addresses your concern @SchSeba

File: cmd/sriov/main.go
Comment (wizhaoredhat) [reply cmd/sriov/main.go]: I moved it to packet.go. It seems the most fitting. I also added a docstring as well. Thanks for this comment, it does make it cleaner.

File: pkg/utils/netlink_manager.go
Comment (wizhaoredhat) [reply pkg/utils/netlink_manager.go:1]: In the root folder:
1) make
2) go mod vendor
3) docker run -v "$PWD":/src -w /src vektra/mockery --recursive=true --name=NetlinkManager --output=./pkg/sriov/mocks/ --filename=netlink_manager_mock.go
4) docker run -v "$PWD":/src -w /src vektra/mockery --recursive=true --name=pciUtils --output=./pkg/sriov/mocks/ --filename=pci_utils_mock.go --exported

I updated the files with the autogenerated files from mockery. I ran "sudo make test" and it seems everything is working properly.

Comment (Eoghan1232) [reply pkg/utils/netlink_manager.go:1]: perhaps some doc put under the "mocks" dir  ?

Comment (adrianchiris) [reply pkg/utils/netlink_manager.go:1]: afaik, its a common practice to put mocked interfaces of a package under `mocks` folder in the same location of the package itself.

this can be addressed as a followup by creating a `Developer.md` with a section describing this.
imo it can be toplevel doc next to `README.md` WDYT ?

File: pkg/sriov/mocks/netlink_manager_mock.go
Comment (adrianchiris) [inline pkg/sriov/mocks/netlink_manager_mock.go:2]: netlink manager interface and impl was moved to utils package

can you move this mock to be under `pkg/utils/mocks` ?

File: pkg/utils/netlink_manager.go
Comment (wizhaoredhat) [reply pkg/utils/netlink_manager.go:1]: I think its a good idea Adrian and Eoghan. I put this document "Developer.md" at the root of the project directory.

File: pkg/sriov/mocks/netlink_manager_mock.go
Comment (wizhaoredhat) [reply pkg/sriov/mocks/netlink_manager_mock.go]: Good catch, I moved the mock file accordingly.

**Review by adrianchiris — APPROVED** [review]
lets wait for CI to pass, otherwise LGTM

thanks for addressing my comments !

**Comment by wizhaoredhat** [comment]
I am not sure why the Mellanox CI is failing when we only changed the test file and mocks... Any idea @adrianchiris ?

**Comment by adrianchiris** [comment]
it should run properly now

**Comment by wizhaoredhat** [comment]
> it should run properly now

Great, could we merge this pull request now?

**Review by Eoghan1232 — APPROVED** [review]
lgtm

**Comment by Eoghan1232** [comment]
Merging now.

**Pull request merged**
